squire_sdk/src/api/request.rs

```diff
@@ -0,0 +1,146 @@
+#[allow(unused_imports)]
+use std::convert::Infallible;
+
+#[cfg(feature = "server")]
+use axum::{body::HttpBody, handler::Handler, routing::MethodRouter};
+use serde::{de::DeserializeOwned, Serialize};
+
+use super::url::Url;
+#[cfg(feature = "server")]
+use crate::server::state::ServerState;
+
+/// An enum that is used by the `RestRequest` to "know" what `MethodRouter` contructor to call
+pub enum Method {
+    Get,
+    Post,
+    Patch,
+    Delete,
+}
+
+// Only integers, chars, and bools can be used as const generics, so we define consts for each
+// method variant.
+pub const GET: u8 = Method::Get as u8;
+pub const POST: u8 = Method::Post as u8;
+pub const PATCH: u8 = Method::Patch as u8;
+pub const DELETE: u8 = Method::Delete as u8;
+
+/// A function used to turn const-generics into Methods at compile time.
+///
+/// NOTE: This panics only at compile time, but it will do so when you run `cargo build` not `cargo
+/// check`.
+const fn to_method<const M: u8>() -> Method {
+    match M {
+        GET => Method::Get,
+        POST => Method::Post,
+        PATCH => Method::Patch,
+        DELETE => Method::Delete,
+        _ => panic!("Invalid method value"),
+    }
+}
+
+/// A trait that abstracts the connections between the data used in API definitions on the server's
+/// side. This trait is not intended to be implmented directly. Instead, a type should implement
+/// one of the method-specific traits since each of those has a blanket implementation for this
+/// trait.
+// TODO: Remove the generic `N` here. This trait does not need to have a full URL, only the string.
+pub trait RestRequest<const N: usize, const M: u8>: Serialize + DeserializeOwned {
+    const ROUTE: Url<N>;
+    const METHOD: Method = to_method::<M>();
```

Comment:
You are partially correct. Let's say we implement `RestRequest<0, GET>` for some type `T` and then implement `RestRequest<0, POST>` for the same `T`. If you want to get the method for `T`'s `RestRequest`, you need to specify the `M` value. This allows you to specify both routes via `<T as RestRequest<0, GET>>::ROUTE` and similarly for the POST implementation.

Until ideal conditions, however, under ideal conditions, this trait wouldn't have any const-generics. They are needed because the following code is not allowed (free generic parameters):
```rust
pub trait RestRequest: Serialize + DeserializeOwned {
    const ROUTE: &'static str;
    const METHOD: Method;
}

impl<const N: usize, const M: u8, T> RestRequest for T
where T: GetRequest<N, M> { ... }
```
I believe this will be part of a future feature: either const generic-associated types or the rest of the const-generics feature. For now, I think this is the best we have.

One last thing, we can't reference `Self::` in const contexts, so we are unable to somehow implement traits like `GetRequest` from `RestRequest` using only associated constants inside `RestRequest`.